scanners/sitemap.py

```diff
@@ -16,57 +29,70 @@
 
 # Required scan function. This is the meat of the scanner, where things
 # that use the network or are otherwise expensive would go.
-#
 # Runs locally or in the cloud (Lambda).
 def scan(domain: str, environment: dict, options: dict) -> dict:
     logging.debug("Scan function called with options: %s" % options)
 
-    results = {}
+    sitemap = None
+    fqd = "https://%s" % domain  # note lack of trailing slash
+    results = {
+        'status_code': None,
+        'final_url': None,
+        'url_tag_count': None,
+        'pdfs_in_urls': None,
+        'sitemap_locations_from_index': [],
+        'robots': None,
+        'crawl_delay': None,
+        'sitemap_locations_from_robotstxt': []
+    }
 
     # get status_code and final_url for sitemap.xml
     try:
-        response = requests.head("https://" + domain + '/sitemap.xml', allow_redirects=True, timeout=4)
-        results['status_code'] = str(response.status_code)
-        results['final_url'] = response.url
-    except Exception:
-        logging.debug("could not get data from %s/sitemap.xml", domain)
-        results['status_code'] = str(-1)
-        results['final_url'] = ''
+        sitemap = requests.get(fqd + '/sitemap.xml', timeout=4)
+        results['status_code'] = sitemap.status_code
+        results['final_url'] = sitemap.url
+    except Exception as error:
+        results['status_code'] = "Could not get data from %s/sitemap.xml: %s" % (domain, error)
 
-    # search sitemap and count the <url> tags
-    url = 'https://' + domain + '/sitemap.xml'
-    i = 0
-    try:
-        with urllib.request.urlopen(url, timeout=5) as sitemap:
-            for _, element in etree.iterparse(sitemap):
-                tag = etree.QName(element.tag).localname
-                if tag == 'url':
-                    i = i + 1
-                element.clear()
-    except Exception:
-        logging.debug('error while trying to retrieve sitemap.xml')
-    results['url_tag_count'] = i
+    # Check once more that we have a usable sitemap before parsing it
+    if sitemap and sitemap.status_code == HTTPStatus.OK:
+        soup = BeautifulSoup(sitemap.text, 'lxml')
+        urls = soup.find_all('url')
+        results['url_tag_count'] = len(urls)
+        # and how many of those URLs appear to be PDFs
+        if urls:
+            results['pdfs_in_urls'] = len([u for u in urls if '.pdf' in u.get_text()])
```

Comment:
Any reason not to check file extension with `endswith`? I think it might be slightly more accurate.

Reply:
No strong opinion. But it's also another case where 100% accuracy isn't needed, so whatever's simplest (and fastest) is probably the way to go.